### PR TITLE
Refactor unlocking logic

### LINKED ISSUE #322: Overworld Backend: Refactor Unlocking Logic
## Current Behavior
The unlocking logic works, but needs some improvements.

## Expected Behavior
All areas are checked for if the lecturer configures a new area.
The progression doesn't reset.

## DoD
- [ ] All areas are checked for if they are unlocked at the login of a user.

### PR DESCRIPTION
**Pull request by Aaron2000119**
Closes https://github.com/Gamify-IT/issues/issues/322.

With this PR, areas can get unlocked even if they are added after the initial release of a course. The either get added if the player has completed the furthest area or has already progressed past the newly unlocked area.

### CODE REVIEW

File: src/main/java/de/unistuttgart/overworldbackend/service/MinigameTaskService.java [hunk 1]
```diff
@@ -177,6 +182,37 @@ private void minigameAdded(final int courseId, final int worldIndex, final Optio
...
+     * @param area Area which might get unlocked
+     * @param playerStatistic statistic of the player
+     */
+    private void unlockAreaIfProgressFurther(final Area area, final PlayerStatistic playerStatistic) {
+        final AreaComparator areaComparator = new AreaComparator();
+        final Area furthestPlayerArea = playerStatistic
+            .getUnlockedAreas()
+            .stream()
+            .max(areaComparator)
+            .orElseThrow(() -> new ResponseStatusException(HttpStatus.BAD_REQUEST, "PlayerStatistic not initiallized"));
```
Comment (maexled) [inline src/main/java/de/unistuttgart/overworldbackend/service/MinigameTaskService.java]: add exception in java doc

File: src/main/java/de/unistuttgart/overworldbackend/service/MinigameTaskService.java [hunk 2]
```diff
@@ -177,6 +182,37 @@ private void minigameAdded(final int courseId, final int worldIndex, final Optio
...
+        if (areaComparator.compare(furthestPlayerArea, area) > 0) {
+            playerStatistic.getUnlockedAreas().add(area);
+        } else {
+            playerStatisticService.checkForUnlockedAreas(furthestPlayerArea, playerStatistic);
```
Comment (maexled) [inline src/main/java/de/unistuttgart/overworldbackend/service/MinigameTaskService.java:216]: Why is this needed?

File: src/main/java/de/unistuttgart/overworldbackend/service/MinigameTaskStatisticService.java [hunk 3]
```diff
@@ -139,6 +139,7 @@ private List<PlayerTaskActionLog> getOrderedPlayerTaskActionLogByDate(
         return successfulPlayerTaskStatistic
             .getPlayerTaskActionLogs()
             .stream()
+            .filter(e -> e.getDate() != null)
```
Comment (maexled) [inline src/main/java/de/unistuttgart/overworldbackend/service/MinigameTaskStatisticService.java]: In commit https://github.com/Gamify-IT/overworld-backend/pull/68/commits/19d0ed63db522cd2808332bdbb9d71bdeee9f4f8 i figured out, that the date value is sometimes null, because the annotation @CreationTimestamp is not working as excepted. I fixed it with using @PrePersist and set the date to current.

File: src/main/java/de/unistuttgart/overworldbackend/service/PlayerStatisticService.java [hunk 4]
```diff
@@ -61,6 +61,18 @@ public PlayerStatistic getPlayerStatisticFromCourse(final int courseId, final St
...
     }
 
+    /**
+     * get statistics from a player course
+     *
+     * @param courseId the id of the course
+     * @param userId   the playerId of the player searching for
+     * @return the found playerstatistic
+     * @throws ResponseStatusException (404) when playerstatistic with courseId and userId could not be found
+     */
```
Comment (maexled) [inline src/main/java/de/unistuttgart/overworldbackend/service/PlayerStatisticService.java]:
```suggestion
    /**
     * Get statistics from a players in a course.
     *
     * @param courseId the id of the course
     * @return the found playerstatistics
     */
```
There is no exception thrown

**Review by maexled — COMMENTED** [review]
Overall looks good. Do not forget to change the image tags back.

File: src/main/java/de/unistuttgart/overworldbackend/service/MinigameTaskService.java (hunk 2 above)
Comment (Aaron2000119) [reply src/main/java/de/unistuttgart/overworldbackend/service/MinigameTaskService.java:216]: if a new world gets added which is further in the progression than the current unlocked worlds and the player has everything unlocked already and also 100% progression in the furtherst area, we need to unlock the newly added area

**Review by maexled — APPROVED** [review]
lgtm
